modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+                        minStatus = base - 1 + FORWARD_SKIP;
+                    } else {
+                        // since we've already passed the last min status
+                        minStatus = current;
+                    }
+                }
+
+                // same as initialization => no change from gap fill functionality
+                if (base == -1) {
+                    if (maxStatus < current + LARGE_REQUEST_SIZE) {
+                        // signal to switch back to / stay in NORMAL mode
+                        base = current;
+                    } else {
+                        // regular jump forward
+                        base = current > maxRequest ? current : maxRequest;
+                        base += FORWARD_SKIP;
+                    }
+                }
+            }
+
+            if (LOG_SYNC.isDebugEnabled()) {
+                LOG_SYNC.debug(
+                        "min status = {}, max status = {}, max requested = {}, known best = {}, current = {}, returned base = {}",
+                        minStatus,
+                        maxStatus,
+                        knownBest,
+                        maxRequest,
+                        current,
+                        base);
+            }
+
+            // keep track of base
+            if (base > maxRequest) {
+                maxRequest = base;
+            }
+
+            // return new base
+            return base;
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void flush() {
+        lock.writeLock().lock();
+        try {
+            levelSource.flush();
+            queueSource.flush();
+            if (!this.indexSource.isAutoCommitEnabled()) {
+                this.indexSource.commit();
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void close() {
+        lock.writeLock().lock();
+
+        try {
+            levelSource.close();
+        } catch (Exception e) {
+            LOG.error("Not able to close the pending blocks levels database:", e);
+        } finally {
+            try {
+                queueSource.close();
+            } catch (Exception e) {
+                LOG.error("Not able to close the pending blocks queue database:", e);
+            } finally {
+                try {
+                    indexSource.close();
+                } catch (Exception e) {
+                    LOG.error("Not able to close the pending blocks index database:", e);
+                } finally {
+                    lock.writeLock().unlock();
+                }
+            }
+        }
+    }
+
+    private String statusToString() {
+        StringBuilder sb = new StringBuilder("Current status queues:\n");
+        for (QueueInfo i : status.values()) {
+            sb.append(i.toString());
+            sb.append('\n');
+        }
+        return sb.toString();
+    }
+
+    static class QueueInfo {
```

Comment:
I think this can be private

Reply:
I added the `@VisibleForTesting` tag in 2bc2919 since one of the testing methods returns this type of object